mono/tests/exception20.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.IO;
+
+class CustomException : Exception
+{
+	public int Value;
+	public CustomException (int value) => this.Value = value;
+}
+
+class C
+{
+	static CustomException e;
+
+	static void Throw ()
+	{
+		try {
+			throw new CustomException(1);
+		} catch (CustomException ex) {
+			e = ex;
+		}
+	}
+
+	static int FrameCount (Exception ex)
+	{
+			string fullTrace = ex.StackTrace;
+			if (fullTrace == null)
+				throw new Exception ("Empty trace found!");
+
+			string[] frames = fullTrace.Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
+
+			// Ignore metadata
+			frames = frames.Where (l => !l.StartsWith ("[")).ToArray ();
+
+			return frames.Length;
+	}
+
+	// throws on false anyway
+	[MethodImpl(MethodImplOptions.NoInlining)]
+	static bool CheckTrace (Exception ex, int numFramesToExpect, bool whatToReturn = true)
+	{
+		int frames = FrameCount (ex);
+		if (frames != numFramesToExpect) {
+			throw new Exception ($"Exception carried {frames} frames along with it when it should have reported {numFramesToExpect}. Full trace:\n---\n{ex.StackTrace}\n---");
+		}
+
+		return whatToReturn;
+	}
+
+	[MethodImpl(MethodImplOptions.NoInlining)]
+	static void ThrowCustomExceptionWithValueButFailsToCatchWithFilter ()
+	{
+		try {
+			ThrowCustomException (1);
+		} catch (CustomException ce) when (ce.Value == 9999) {
+			throw new Exception ("This should NEVER be hit");
+		}
+	}
+
+	[MethodImpl(MethodImplOptions.NoInlining)]
+	static void ThrowCustomException (int value)
+	{
+		throw new CustomException (value);
+	}
+
+	[MethodImpl(MethodImplOptions.NoInlining)]
+	static void ThrowFileNotFoundException ()
+	{
+		throw new FileNotFoundException ();
+	}
+
+	public static void Main ()
+	{
+		Throw ();
+
+		// Single frame
+		// Filter returns false
+		try {
+			throw e;
```

Comment:
why do you reuse the exception here? Also, you should mark `Throw ()` with `[MethodImpl(MethodImplOptions.NoInlining)]`, otherwise the framecount is 1 too very likely.

Reply:
That's a good question. This was based on `exception19.cs` where the exception was being reused too. I think maybe we can enhance this to test that the trace was indeed reset when the filter invoked. Currently, trace from both the original exception and the new trace here would have 1 frame, but if we change `Throw` to instead have two frames then we will be checking if the filter got a *new* trace or not. How does that sound?

```
+		} catch (Exception ex) when (CheckTrace (ex, 2, true)) {
+			CheckTrace (ex, 2);
+		}
```

Reply:
sounds good 👍